### PR TITLE
Initial cut on GPS IMU orientation compensation

### PR DESCRIPTION
**Pull request by digitalentity**
- [x] Implement ACC/GPS error correction according to the paper by Bill Premerlani (Roll-Pitch Gyro Drift Compensation) [1]
- [x] Create interface code between GPS and IMU
- [ ] Test that error vector is calculated correctly
- [ ] Flight test

[1] - [RollPitchDriftCompensationRv3.pdf](https://github.com/iNavFlight/inav/files/2866638/RollPitchDriftCompensationRv3.pdf)
[2] - [CentrifugalCompensation.pdf](https://github.com/iNavFlight/inav/files/2949977/CentrifugalCompensation.pdf)

### CODE REVIEW

**Comment by hamishfagg** [comment]
Not sure if you've seen this but just an FYI - The same guy later posted this about doing the same thing without GPS:

https://diydrones.com/forum/topics/centrifugal-compensation-with-only-gyros-and-accelerometers

**Comment by digitalentity** [comment]
Studying the Premerlani's paper on centrifugal compensation w/o GPS and airspeed. It makes sense actually...

File: src/main/flight/imu.c [hunk 1]
```diff
@@ -525,12 +606,18 @@ static void imuCalculateEstimatedAttitude(float dT)
     }
 #endif
 
+    bool useGPSVelEF = sensors(SENSOR_GPS) && STATE(GPS_FIX) && gpsSol.flags.validVelNE && gpsSol.flags.validVelNE;
```
Comment (giacomo892) [inline src/main/flight/imu.c]: gpsSol.flags.validVelNE && gpsSol.flags.validVelNE; 
Should be gpsSol.flags.validVelNE && gpsSol.flags.validVelD;

Comment (digitalentity) [reply src/main/flight/imu.c]: Nice catch!

**Comment by hamishfagg** [comment]
@digitalentity Do you need testing done of this? I don't have a rig set up but I have a spare 411-wing and foam board..

**Comment by digitalentity** [comment]
@IVData this doesn't work at the moment. I'm working on figuring out what's wrong

File: src/main/flight/imu.c [hunk 2]
```diff
@@ -377,14 +385,87 @@ static void imuMahonyAHRSupdate(float dt, const fpVector3_t * gyroBF, const fpVe
...
+                vectorScale(&vTmp1, &vTmp1, 1.0f / vVelEF_integralTime);
+
+                // Calculate Ge - (V2 - V1) / (T2 - T1)
+                vectorSub(&vTmp1, &vGravity, &vTmp1);
+                const float vTmp1Length = sqrtf(vectorNormSquared(&vTmp1));
+
+                if (vTmp1Length > 0.01f) {
+                    // Calculate acceleration by taking a derivative of acceleration integral
+                    // This effectively calculates average acceleration, but in a way that's more immune to jitter
+                    vectorScale(&vTmp2, &vVelEF_AccIntegal, 1.0f / vVelEF_integralTime);
```
Comment (giacomo892) [inline src/main/flight/imu.c]: There is something wrong there, vTmp2 which should have average acceleration between GPS updates in cm/s/s seems like to be multiplied by a factor of 1000

File: src/main/flight/imu.c [hunk 3]
```diff
@@ -377,14 +385,87 @@ static void imuMahonyAHRSupdate(float dt, const fpVector3_t * gyroBF, const fpVe
...
+            if (velEFNew) {
+                fpVector3_t vTmp1, vTmp2;
+
+                // Calculate (V2 - V1) / (T2 - T1)
+                // Here T2-T1 = vVelEF_integralTime (accumulated time between two GPS updates, also time window we used to integrate accelerometer)
+                vectorSub(&vTmp1, &vVelEF_prev, velEF);
+                vectorScale(&vTmp1, &vTmp1, 1.0f / vVelEF_integralTime);
+
+                // Calculate Ge - (V2 - V1) / (T2 - T1)
+                vectorSub(&vTmp1, &vGravity, &vTmp1);
```
Comment (giacomo892) [inline src/main/flight/imu.c]: Ge is the gravity vector expected in EF (NED) which is 0 0 1 (1G down) so it is somehow normalized.
When we subtract the acceleration as calculated from GPS, that has a different unit and scale so there is something wrong here.

File: src/main/flight/imu.c [hunk 4]
```diff
@@ -403,24 +403,39 @@ static void imuMahonyAHRSupdate(float dt, const fpVector3_t * gyroBF, const fpVe
         if (velEF && vVelEF_initialized) {
             // We have valid velocity from GPS, use that to calculate correction vector
             fpVector3_t vAccEF;
+            vectorZero(&vAccEF);
 
             // Rotate ACC from BF to EF and accumulate
             quaternionRotateVectorInv(&vAccEF, accBF, &orientation);
+
+            vectorScale(&vAccEF, &vAccEF, 1.0f/980.0f); //scale g
+            vectorScale(&vAccEF, &vAccEF, dt); //integrate
```
Comment (digitalentity) [inline src/main/flight/imu.c]: This should be it. Damn, how could I miss integration step?

File: src/main/flight/imu.c [hunk 5]
```diff
@@ -403,24 +403,39 @@ static void imuMahonyAHRSupdate(float dt, const fpVector3_t * gyroBF, const fpVe
         if (velEF && vVelEF_initialized) {
             // We have valid velocity from GPS, use that to calculate correction vector
             fpVector3_t vAccEF;
+            vectorZero(&vAccEF);
```
Comment (digitalentity) [inline src/main/flight/imu.c]: This should be redundant, `quaternionRotateVectorInv` will overwrite it anyway

File: src/main/flight/imu.c [hunk 6]
```diff
@@ -403,24 +403,39 @@ static void imuMahonyAHRSupdate(float dt, const fpVector3_t * gyroBF, const fpVe
...
+            vectorScale(&vAccEF, &vAccEF, dt); //integrate 
+
             vectorAdd(&vVelEF_AccIntegal, &vVelEF_AccIntegal, &vAccEF);
             vVelEF_integralTime += dt;
 
             // Now if we got a GPS update we can calculate error vector according to
             // the paper by Bill Premerlani (Roll-Pitch Gyro Drift Compensation)
             if (velEFNew) {
                 fpVector3_t vTmp1, vTmp2;
+                vectorZero(&vTmp1);
```
Comment (digitalentity) [inline src/main/flight/imu.c]: Same, no need to zero tmp vars

File: src/main/flight/imu.c [hunk 7]
```diff
@@ -403,24 +403,39 @@ static void imuMahonyAHRSupdate(float dt, const fpVector3_t * gyroBF, const fpVe
...
 
                 // Calculate (V2 - V1) / (T2 - T1)
                 // Here T2-T1 = vVelEF_integralTime (accumulated time between two GPS updates, also time window we used to integrate accelerometer)
-                vectorSub(&vTmp1, &vVelEF_prev, velEF);
+             
+                fpVector3_t velEFLocal;
+
+                vectorScale(&velEFLocal, velEF, 1.0f);
+
+                vectorScale(&velEFLocal, &velEFLocal, 1.0f / 980.0f); //scale EF to match g scale
```
Comment (digitalentity) [inline src/main/flight/imu.c]: I don't think we should scale to G's, cm/s/s should be enough as long as units are the same for GPS-based acc and IMU-based

**Comment by stronnag** [comment]
I tried this on a tricopter, MATEKF405. Unfortunately, I forgot that the SD Card has been in a 'plane for the last 12 months ....

It did not go well (appreciate that without a blackbox log, this is worthless speculation).

* Started off with PH, worked OK
* Flew a WP mission
* Some way onto the mission the craft started to roll to starboard. A smooth and continuous roll
* The craft took up a strange attitude as it tried to maintain heading whilst following a curved track to the right due to the roll.
* Took "control", sort of landed, breaking 2 props. 

I do have a "mwp geo" log which has no useful FC state, it does show what happened, craft should be flying directly along the dashed orange track in WP mode.
tri-crahs

And the AH looked as below (machine should have been straight and level,  very light wind.
Screenshot-20200122202611-154x143

**Comment by giacomo892** [comment]
Thanks for the test @stronnag , i'm sorry for the props.
Continuing to work on this

**Comment by avsaase** [comment]
I'm going through the second paper and while I can follow most of the steps there are a few I don't understand. The way the vectors are defined also confuse me. Other than that this seems totatlly doable so if anyone can help me out that would be great.

**Comment by nullie** [comment]
I wrote a unit test and I found a possible issue: after conversion of IMU orientation to quaternion, "earth frame" in IMU has Y axis pointing west instead of east, you can see that in "hacks" in imuTransformVectorEarthToBody and imuTransformVectorBodyToEarth. So, for GPS compensation to work correctly, we have to invert Y of GPS velocity.

Test and fix adapted to this branch are in https://github.com/iNavFlight/inav/pull/7575/files

**Pull request closed without merging**